game-core/src/main/java/games/strategy/engine/data/RouteFinder.java

```diff
@@ -64,7 +65,7 @@ private boolean calculate(final Set<Territory> startSet, final Territory end) {
   private Route getRoute(final Territory start, final Territory destination) {
     final List<Territory> route = new ArrayList<>();
     Territory current = destination;
-    while (current != start) {
```

Comment:
I mentioned that I had a deeper look at this class in ron's PR.
Technically current should never be null (even if this explicitly checks for null).
getRoute is a private method that's only ever being called from findRoute when it returns true i.e. a valid route has been found (it doesn't necessarily has to be feasible for the unit, but the route is possible). Therefore going back from the destination to the previous territories should always result in getting back to the start.
If we'd ever be reaching null here this could only "possibly" be (haven't actually tried to do so, just speculating) when findRoute is being passed null as start territory or something.
By checking preconditions in findRoute those kinds of problems could be avoided.


But don't try to do the class correctly.
I really want to implement a proper breadth-first search (which could possibly be extended to a dijkstra algorithm to address the comment before the class definition) in a week, when I have time for TripleA again. 😅

Reply:
> I really want to implement a proper breadth-first search in a week

Yeah, I saw your comment in Ron's PR.  That's why I got this fix in quick, so I wouldn't be tempted to address it while you were possibly in the middle of a re-write. :smile:

Reply:
I'll leave it up to you if you want to change this into a regular equals then.

Reply:
Ok, following the code upstream, I now see that `start` will never be `null` due to precondition checks.  So I switched it to `Object#equals()` with `start` as the receiver to guarantee no NPE.  baf1bb3bd